File: golang/815.bus-routes.go
func numBusesToDestination(routes [][]int, source int, target int) int {
    // no need to take a bus
    if source == target {
        return 0
    }

    gr := make(map[int][]int)
    stops := 0
    for bus, route := range routes {
        for _, stop := range route {
            gr[stop] = append(gr[stop], bus)
            stops++
        }
    }
    // prelimirary check: if source or target are not on any bus line
    // then it's impossible to get from source to target
    if len(gr[source]) == 0 || len(gr[target]) == 0 {
        return -1
    }

    // queue of tuples (stop, distance)
    // with reserve capacity number of stops
    q := make([][2]int, 0, stops)
    q = append(q, [2]int{source, 0})
    // visited buses and visited stops
    vis_stops := make(map[int]bool)
    vis_buses := make([]bool, len(routes))

    for len(q) > 0 {
        node := q[0]
        q = q[1:]
        stop, dist := node[0], node[1]

        if stop == target {
            return dist
        }

        for _, bus := range gr[stop] {
            if vis_buses[bus] {
                continue
            }
            vis_buses[bus] = true

            for _, next_stop := range routes[bus] {
                if vis_stops[next_stop] {
                    continue
                }
                vis_stops[next_stop] = true

                q = append(q, [2]int{next_stop, dist + 1})
            }
        }
    }

    return -1
}

File: python/815.bus-routes.py
class Solution:
    def numBusesToDestination(self, routes: List[List[int]], source: int, target: int) -> int:
        # no need to take bus at all
        if source == target:
            return 0

        # build graph of stop -> set of buses
        gr = defaultdict(set)
        for bus, route in enumerate(routes):
            for stop in route:
                gr[stop].add(bus)
        
        queue = deque([(source, 0)])
        vis_stops = set()
        vis_buses = set()

        while queue:
            stop, dist = queue.popleft()

            if stop == target:
                return dist

            for bus in gr[stop]:
                if bus in vis_buses:
                    continue
                vis_buses.add(bus)

                for next_stop in routes[bus]:
                    if next_stop in vis_stops:
                        continue
                    vis_stops.add(next_stop)

                    queue.append((next_stop, dist + 1))

        return -1